### PR TITLE
🌱 Reconcile topology only when necessary

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
The topology controller is responsible to enforce the desired state of the cluster, and it does this ensuring that the spec of the cluster itself and the spec of the dependant objects like e.g. Machine Deployment is consistent with what defined in the ClusterClass.

That means that this controller should not care of changes to objects status (or at least to a subset of changes in status).
This PR introduces two predicates starting to drop reconcile events when the changes are not relevant for the topology controller

/area cluster

### CODE REVIEW

**Comment by fabriziopandini** [comment]
test

File: internal/controllers/topology/cluster/cluster_controller.go [hunk 1]
```diff
@@ -155,6 +165,82 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
...
+			oldObj = dropNotRelevant(oldObj)
+
+			newObj := e.ObjectNew.(*clusterv1.Cluster)
+			if !ok {
+				log.V(4).Info("Expected Cluster", "type", fmt.Sprintf("%T", e.ObjectNew))
+				return false
+			}
+			newObj = dropNotRelevant(newObj)
+
+			return reflect.DeepEqual(oldObj, newObj)
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/controllers/topology/cluster/cluster_controller.go:198]: Additionally. ResourceVersion, managedFields etc will also be different

File: internal/controllers/topology/cluster/cluster_controller.go [hunk 2]
```diff
@@ -155,6 +165,82 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
...
+		},
+		CreateFunc:  func(event.CreateEvent) bool { return true },
+		DeleteFunc:  func(event.DeleteEvent) bool { return true },
+		GenericFunc: func(event.GenericEvent) bool { return false },
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/cluster_controller.go:240]:
```suggestion
		GenericFunc: func(event.GenericEvent) bool { return true },
```

File: internal/controllers/topology/cluster/cluster_controller.go [hunk 3]
```diff
@@ -155,6 +165,82 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
...
+func clusterChangeIsRelevant(scheme *runtime.Scheme, logger logr.Logger) predicate.Funcs {
+	dropNotRelevant := func(cluster *clusterv1.Cluster) *clusterv1.Cluster {
+		c := cluster.DeepCopy()
+		// Drop changes on v1beta2 conditions; when v1beta2 conditions will be moved top level, we will review this
+		// selectively drop changes not relevant for this controller.
+		c.Status.V1Beta2 = nil
+		return c
+	}
+
+	return predicate.Funcs{
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/cluster_controller.go:177]: Let's also add some log level 6 logs

**Comment by sbueringer** [comment]
Did some local testing, all good

/hold cancel

**Pull request merged**